File: shell/services/bunny_shell/notification_center.py
# SPDX-FileCopyrightText: 2026 ComradeArt
# SPDX-License-Identifier: GPL-3.0-or-later
"""Quiet notification center with optional Bunny summary.

GNOME remains the freedesktop daemon. This is Bunny's own history: collapse
repeats, summarize Bunny chatter, keep lock-screen redaction.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping, Sequence

QUIET_DEFAULTS = {
    "quiet": True,
    "max_visible_toasts": 3,
    "max_history": 50,
    "collapse_window_ms": 8000,
    "bunny_summary": True,
    "sensitive_default": True,
}


@dataclass(frozen=True)
class FoldedNotice:
    source: str
    title: str
    body: str
    severity: str
    at: int
    count: int
    sensitive: bool


def _key(source: str, title: str) -> str:
    return f"{source}\n{title}"


def fold_notifications(
    items: Sequence[Mapping[str, Any]] | None,
    *,
    quiet: bool = True,
    bunny_summary: bool = True,
    now: int = 0,
) -> tuple[FoldedNotice, ...]:
    kept: list[FoldedNotice] = []
    seen: dict[str, int] = {}
    for raw in items or ():
        if not isinstance(raw, Mapping):
            continue
        severity = str(raw.get("severity") or "info")
        source = str(raw.get("source") or "Bunny")
        title = str(raw.get("title") or raw.get("body") or "Notification")
        body = str(raw.get("body") or "")
        stamp = int(raw.get("at") or now or 0)
        key = _key(source, title)
        if quiet and severity == "info":
            previous = seen.get(key)
            if previous is not None and stamp - previous <= QUIET_DEFAULTS["collapse_window_ms"] and kept:
                last = kept[-1]
                if _key(last.source, last.title.split(" (")[0]) == key or _key(last.source, last.title) == key:
                    count = last.count + 1
                    kept[-1] = FoldedNotice(
                        source=last.source,
                        title=f"{title} ({count})",
                        body=f"{title} ({count})" if bunny_summary and source == "Bunny" else last.body,
                        severity=last.severity,
                        at=last.at,
                        count=count,
                        sensitive=last.sensitive,
                    )
                    continue
            seen[key] = stamp
        kept.append(
            FoldedNotice(
                source=source,
                title=title,
                body=body,
                severity=severity,
                at=stamp,
                count=1,
                sensitive=raw.get("sensitive", True) is not False,
            )
        )
    return tuple(kept[-QUIET_DEFAULTS["max_history"]:])


def summarize_bunny(items: Sequence[FoldedNotice]) -> str | None:
    bunny = [item for item in items if item.source == "Bunny"]
    if not bunny:
        return None
    if len(bunny) == 1:
        return bunny[0].title
    errors = sum(1 for item in bunny if item.severity == "error")
    if errors:
        return f"Bunny: {len(bunny)} updates, {errors} need attention"
    return f"Bunny: {len(bunny)} updates"


def build_notification_center(
    items: Sequence[Mapping[str, Any]] | None = None,
    *,
    quiet: bool = True,
    bunny_summary: bool = True,
    do_not_disturb: bool = False,
) -> dict[str, Any]:
    folded = () if do_not_disturb else fold_notifications(
        items, quiet=quiet, bunny_summary=bunny_summary
    )
    return {
        "kind": "NotificationCenter",
        "quiet": quiet,
        "bunnySummary": bunny_summary,
        "doNotDisturb": do_not_disturb,
        "maxVisibleToasts": QUIET_DEFAULTS["max_visible_toasts"],
        "summary": summarize_bunny(folded) if bunny_summary else None,
        "items": [
            {"title": item.title, "body": item.body, "severity": item.severity, "source": item.source}
            for item in folded
        ],
        "emptyCopy": (
            "Do Not Disturb is on. Bunny is not showing notices."
            if do_not_disturb
            else "No notices. Bunny stays quiet unless something needs you."
        ),
        "spam": False,
        "companionRequired": False,
        "accessibleName": "Notification center",
    }


def should_toast(
    level: str,
    message: str,
    recent: Sequence[Mapping[str, Any]] | None = None,
    *,
    quiet: bool = True,
    now: int = 0,
) -> bool:
    if not quiet:
        return True
    if level in {"error", "warning"}:
        return True
    text = str(message or "")
    window = QUIET_DEFAULTS["collapse_window_ms"]
    for entry in recent or ():
        if (
            entry.get("level") == "info"
            and entry.get("message") == text
            and now - int(entry.get("at") or 0) < window
        ):
            return False
    return True

File: shell/components/gnome-shell-extension/lib/notificationCenter.js
// SPDX-FileCopyrightText: 2026 ComradeArt
// SPDX-License-Identifier: GPL-3.0-or-later
//
// Notification center: quiet defaults, optional Bunny summary, no spam.
// GNOME still owns freedesktop delivery. This is Bunny's own tray plus a
// bounded history the GTK surface can also render.

import {buildNotification} from './design/primitives.js';
import {SCREEN_QUESTIONS} from './design/tokens.js';

export const NOTIFICATION_QUIET_DEFAULTS = Object.freeze({
    quiet: true,
    maxVisibleToasts: 3,
    maxHistory: 50,
    collapseWindowMs: 8000,
    bunnySummary: true,
    sensitiveDefault: true,
});

function keyFor(item) {
    return `${String(item.source || 'Bunny')}\n${String(item.title || item.body || '')}`;
}

/**
 * Collapse repeats and info-level chatter. Errors still get their own row.
 */
export function foldNotifications(items, {
    quiet = true, bunnySummary = true, now = 0,
} = {}) {
    const rows = Array.isArray(items) ? items : [];
    const kept = [];
    const seen = new Map();
    for (const raw of rows) {
        if (!raw || typeof raw !== 'object')
            continue;
        const severity = String(raw.severity || 'info');
        const source = String(raw.source || 'Bunny');
        const title = String(raw.title || raw.body || 'Notification');
        const body = String(raw.body || '');
        const stamp = Number(raw.at || now) || 0;
        const key = keyFor({source, title});
        if (quiet && severity === 'info') {
            const previous = seen.get(key);
            if (previous !== undefined && stamp - previous <= NOTIFICATION_QUIET_DEFAULTS.collapseWindowMs) {
                const last = kept[kept.length - 1];
                if (last && keyFor(last) === key) {
                    last.count = (last.count || 1) + 1;
                    last.body = bunnySummary && source === 'Bunny'
                        ? `${title} (${last.count})`
                        : last.body;
                    continue;
                }
            }
            seen.set(key, stamp);
        }
        kept.push({
            source,
            title,
            body,
            severity,
            at: stamp,
            count: 1,
            sensitive: raw.sensitive !== false,
        });
    }
    return kept;
}

export function summarizeBunny(items) {
    const bunny = (Array.isArray(items) ? items : []).filter(
        item => String(item?.source || 'Bunny') === 'Bunny');
    if (bunny.length <= 1)
        return bunny[0] ? buildNotification({
            title: bunny[0].title, body: bunny[0].body, severity: bunny[0].severity || 'info',
        }) : null;
    const errors = bunny.filter(item => item.severity === 'error').length;
    const title = errors
        ? `Bunny: ${bunny.length} updates, ${errors} need attention`
        : `Bunny: ${bunny.length} updates`;
    return buildNotification({
        title,
        body: 'Open the notification center for the rest. Bunny will not ping for each step.',
        severity: errors ? 'warning' : 'info',
    });
}

export function buildNotificationCenter({
    items = [], quiet = true, bunnySummary = true, doNotDisturb = false,
} = {}) {
    const folded = doNotDisturb ? [] : foldNotifications(items, {quiet, bunnySummary});
    const summary = bunnySummary ? summarizeBunny(folded) : null;
    const visible = folded.slice(-NOTIFICATION_QUIET_DEFAULTS.maxHistory);
    return {
        kind: 'NotificationCenter',
        quiet,
        bunnySummary,
        doNotDisturb: Boolean(doNotDisturb),
        maxVisibleToasts: NOTIFICATION_QUIET_DEFAULTS.maxVisibleToasts,
        summary,
        items: visible.map(item => buildNotification({
            title: item.count > 1 ? `${item.title} (${item.count})` : item.title,
            body: item.body,
            severity: item.severity,
        })),
        emptyCopy: doNotDisturb
            ? 'Do Not Disturb is on. Bunny is not showing notices.'
            : 'No notices. Bunny stays quiet unless something needs you.',
        spam: false,
        companionRequired: false,
        questions: SCREEN_QUESTIONS,
        accessibleName: 'Notification center',
        styleClass: 'bunny-notification-center',
        canFocus: true,
    };
}

/**
 * Whether a live toast should appear. Quiet defaults drop duplicate info.
 */
export function shouldToast(level, message, recent = [], {quiet = true, now = Date.now()} = {}) {
    if (!quiet)
        return true;
    if (level === 'error' || level === 'warning')
        return true;
    const text = String(message ?? '');
    return !recent.some(entry =>
        entry.level === 'info'
        && entry.message === text
        && now - Number(entry.at || 0) < NOTIFICATION_QUIET_DEFAULTS.collapseWindowMs);
}
